tower-http/src/catch_panic.rs

```diff
@@ -0,0 +1,202 @@
+        CatchPanicLayer::new()
+    }
+}
+
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for CatchPanic<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    ResBody: Body<Data = Bytes> + Send + 'static,
+{
+    type Response = Response<UnsyncBoxBody<Bytes, ResBody::Error>>;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        match std::panic::catch_unwind(AssertUnwindSafe(|| self.inner.call(req))) {
+            Ok(future) => ResponseFuture {
+                kind: Kind::Future {
+                    future: AssertUnwindSafe(future).catch_unwind(),
+                },
+            },
+            Err(panic_err) => ResponseFuture {
+                kind: Kind::Panicked {
+                    panic_err: Some(panic_err),
+                },
+            },
+        }
+    }
+}
+
+pin_project! {
+    /// Response future for [`CatchPanic`].
+    pub struct ResponseFuture<F> {
+        #[pin]
+        kind: Kind<F>,
+    }
+}
+
+pin_project! {
+    #[project = KindProj]
+    enum Kind<F> {
+        Panicked {
+            panic_err: Option<Box<dyn Any + Send + 'static>>,
+        },
+        Future {
+            #[pin]
+            future: CatchUnwind<AssertUnwindSafe<F>>,
+        }
+    }
+}
+
+impl<F, ResBody, E> Future for ResponseFuture<F>
+where
+    F: Future<Output = Result<Response<ResBody>, E>>,
+    ResBody: Body<Data = Bytes> + Send + 'static,
+{
+    type Output = Result<Response<UnsyncBoxBody<Bytes, ResBody::Error>>, E>;
+
+    #[allow(warnings)]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.project().kind.project() {
+            KindProj::Panicked { panic_err } => Poll::Ready(Ok(panic_err_to_response(
+                panic_err.take().expect("future polled after completion"),
+            ))),
+            KindProj::Future { future } => match ready!(future.poll(cx)) {
+                Ok(Ok(res)) => Poll::Ready(Ok(res.map(|body| body.boxed_unsync()))),
+                Ok(Err(svc_err)) => Poll::Ready(Err(svc_err)),
+                Err(panic_err) => Poll::Ready(Ok(panic_err_to_response(panic_err))),
+            },
+        }
+    }
+}
+
+fn panic_err_to_response<E>(
+    err: Box<dyn Any + Send + 'static>,
+) -> Response<UnsyncBoxBody<Bytes, E>> {
+    let message: Cow<_> = if let Some(s) = err.downcast_ref::<String>() {
+        s.to_owned().into()
+    } else if let Some(s) = err.downcast_ref::<&str>() {
+        (*s).into()
+    } else {
+        "`CatchPanic` middleware was unable to obtain panic info".into()
+    };
+    let body = format!("Service panicked: {}", message);
+
+    let mut res = Response::new(Full::from(body).map_err(|err| match err {}).boxed_unsync());
```

Comment:
I dont think we should hand out the panic message by default to the user. Instead, we should probably `error!` log the exception and just return `500 Internal Server Error` with an empty response - but leave the exact behaviour up to customization as you suggested.

Rationale: I would think it's a bit unprofessional if I GET from some SaaS service and and get back a 500 with the body ~~`undefined is not a function`~~`index out of bounds: the len is 0 but the index is 1`. Also, the current approach does not leave any trace of the panic in the logs.

Reply:
yeah, I think it might be desirable to have some way to configure the middleware with whether or not it actually exposes the panic message in the response. you might want to turn it on for local development, but disable it in production, both for making the app look nice, and also potentially for security reasons (you maybe don't want to expose a panic message or backtrace to the entire internet...)

Reply:
I've added support for custom panic handlers now. I made the default just generic an opaque "Service panicked" message and `tracing::error!` the full message. I believe this will work fine for must cases without having too much customization.